Update/add login related mocks.

pachlava opened the pull request:
### Background 
Automating a selected range of UI tests is the project I'm currently part of. There was a foundation in place for the UI tests already, in the form of a number of screen objects, unfinished ScreenshotTest and numerous WireMock files.

First thing I tried is executing the ScreenshotTest (which currently only logs in). It turned out that both mocks and login sequence on UI level are outdated, so even the login was not working.

### Description
I'll provide more detailed comments on the code, but here's the short summary:

1. Used WireMock version has been bumped by Dependabot some time ago with #4025. However, it turns out that this update made the use of WireMock impossible (see wiremock/wiremock/issues/1345 for more). The used version is revered back in current PR.
2. The login flow in the app UI seems to no longer show the magic link screen after entering user email, and now goes directly to the password screen. This PR updates this.
3. A number of mocks were outdated and required a change in order to have the login mocked. Also, one new file was added.

I tried making the changes to the mocks minimal, but enough to have the login 'working'.

### Testing instructions
1. Use the `gradle.properties` from secrets
2. Run `ScreenshotTest` ~> It should now get past the login and reach the `My Store` screen. 

Please note that the test will still be marked as failed.

### Release
These changes do not require release notes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### P.S.
Since this is my first PR in this repo, I might be doing some things wrong. I tried to follow the contribution guidelines, but let me know if there's a thing I should adjust. Also, this is my first experience with mocks - so I'll be happy to hear any constructive feedback on that too.

pachlava committed:
Update login-involved mocks

+ downgrade wiremock back to 2.26.3

pachlava commented on WooCommerce/build.gradle:
```diff
@@ -257,7 +257,7 @@ dependencies {
     // Dependencies for screenshots
     androidTestImplementation 'tools.fastlane:screengrab:2.0.0'
-    androidTestImplementation('com.github.tomakehurst:wiremock:2.27.2') {
+    androidTestImplementation('com.github.tomakehurst:wiremock:2.26.3') {
```
This rollback is caused by [this issue](https://github.com/wiremock/wiremock/issues/1345#issuecomment-656060968).

mokagio replied:
+1 for reverting to get this work over the line 👍 

Looking at WireMock's [tags list](https://github.com/wiremock/wiremock/releases) I can see they have been pushing out quite a few new version since the issue was created. In the future, we should try to upgrade to the latest 🤞

pachlava replied:
Is there a mechanism to prevent the same version bump initiated by dependabot again? Until we have no UI tests in place and attached to CI, such PR will probably be accepted, because there will be no indication of any issues.

mokagio replied:
We should be able to configure that via a [`dependabot.yml` file](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/enabling-and-disabling-version-updates)

pachlava replied:
Thanks! Done in this PR with [this commit](https://github.com/woocommerce/woocommerce-android/pull/4966/commits/5022f50fd9c4849fc7e64b74cd2374774367735c).

pachlava commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/screenshots/login/EmailAddressScreen.kt:
```diff
@@ -11,10 +11,10 @@ class EmailAddressScreen : Screen {
     constructor() : super(EMAIL_ADDRESS_FIELD)
 
-    fun proceedWith(emailAddress: String): MagicLinkScreen {
+    fun proceedWith(emailAddress: String): PasswordScreen {
```
After entering the email address, the user is redirected to PasswordScreen, instead of MagicLinkScreen. This screen still expected the old flow.

pachlava commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/screenshots/ScreenshotTest.kt:
```diff
@@ -38,10 +38,7 @@ class ScreenshotTest : TestBase() {
             .proceedWith(BuildConfig.SCREENSHOTS_URL)
-            // Enter email address
             .proceedWith(BuildConfig.SCREENSHOTS_USERNAME)
-            // No magic link, this is an automated test, enter password manually
-            .proceedWithPassword()
```
Removed the 'proceed' call here, because magic link screen it's not shown any more during the usual login flow.

pachlava commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/screenshots/ScreenshotTest.kt:
```diff
@@ -38,10 +38,7 @@ class ScreenshotTest : TestBase() {
             .selectLogin()
             // Connect a WooCommerce store by URL
             .proceedWith(BuildConfig.SCREENSHOTS_URL)
-            // Enter email address
```
Removed this comment since the call on the next line seems obvious. Let me know if you see sense in keeping it.

mokagio replied:
👍 that's the whole point of so called "self documenting code" 😄 

--- 

From ["Clean Code" by Uncle Bob](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882):

<img width="500" alt="Screen Shot 2019-10-31 at 9 23 28 am" src="https://user-images.githubusercontent.com/1218433/67903657-4a477380-fbc0-11e9-8f6a-e3c3f210f92a.png">

pachlava replied:
This is for sure my favourite book related to software development (not saying I read a lot of others 😅), and it for sure influenced me the most. I found about it thanks to the famous 'The only valid measurement of code quality' picture.

mokagio commented on WooCommerce/src/androidTest/assets/mocks/mappings/me/rest_v11_me_sites.json:
```diff
@@ -10,7 +10,7 @@
             "name": "Automattic Widgets",
             "description": "",
-            "URL": "https:\/\/automatticwidgets.com",
+            "URL": "https:\/\/automatticwidgets.wpcomstaging.com",
```
Have you got some context on this URL change? Or is it something you ended up with by trial and error? (_I have no idea, just curious_ 😄)

pachlava replied:
To be honest, there are still certain blind spots here for me. Chronologically, it was like this:

1. I found out that the mocks are outdated
2. Used Charles to see the requests (hat tip to @rachelmcr 👋), with no WireMock involved. The emulator spoke to the real API.
3. It was very convenient to find out that the same test account with same credentials is really present, but the associated website address contained additional `.wpcomstaging`.
4. While doing a lot of trial and error, I decided to update the website name too, to fully emulate the experience (but in theory, it should not matter).
5. I also noticed that some information (obviously gravatar, but something else too) still gets communicated via real API.
6. Before creating the PR I decided to undo all changes that theoretically should not matter, but reverting the website name resulted in something like `Can not connect to your website` shown after login.

There is still some communication with real API taking place, and this bugs me, because one of the points of using mocks during UI tests is not relying on the potentially unstable/unaccessible API.

mokagio commented on WooCommerce/src/androidTest/assets/mocks/mappings/me/rest_v11_me_sites.json:
```diff
@@ -1,6 +1,6 @@
 {
     "request": {
-        "urlPath": "/rest/v1.1/me/sites/",
+        "urlPathPattern": "/rest/v1.(1|2)/me/sites/",
```
Non-actionable nitpcik: the file name says `_v11_` which I take to be "v1.1" but now we're supporting v1.1 and v1.2. I'm pointing this out, but I don't know whether there's a convention to represent the duality in the file name. I have no prior WireMock experience 😅 Keen to learn more about this together with you 🤓

pachlava replied:
That's a very good point @mokagio ! I renamed both files. I _guess_ that the names are generated automatically during mocks recording, but some files here do not contain this part (even though it's in the request that they mock), probably because they were renamed manually at some point. Initially I thought that even the inner folder structure under `mappings` matters (and same about file names), but it appears to be not the case.

mokagio reviewed (commented):
I'm just going to leave this partial review here while waiting for the tests to build and run, in case I run out of time.

Looking to me, well done 👍 If they work, I think we can merge, but I'd like to hear what @jkmassel (or maybe @rachelmcr if you have time) have to say on the mocks, because I don't know much about them.

pachlava committed:
Rename rest_v11_sites_161477129.json to rest_sites_161477129.json

Since REST v1.2 is used here, the old name made no sense.

pachlava committed:
Rename rest_v11_me_sites.json to rest_me_sites.json

Since REST v1.2 is used here, the old name made no sense.

rachelmcr commented on WooCommerce/src/androidTest/assets/mocks/mappings/me/rest_me_sites.json:
```diff
@@ -10,7 +10,7 @@
             "name": "Automattic Widgets",
             "description": "",
-            "URL": "https:\/\/automatticwidgets.com",
+            "URL": "https:\/\/automatticwidgets.wpcomstaging.com",
```
Just a note that if this data is displayed anywhere where it will end up in screenshots, it might be nice to use a nice-looking URL like `automatticwidgets.com` instead of `automatticwidgets.wpcomstaging.com`. (I'm not sure if it is, but just something to watch out for.)

pachlava replied:
Thanks for taking a look @rachelmcr! Very good point. Literally one minute ago I [answered](https://github.com/woocommerce/woocommerce-android/pull/4966#discussion_r724908045) to Gio about this change 😄. I'll make sure (in this PR) that this does not get displayed onscreen.

rachelmcr reviewed (commented):
I didn't do a full review, but I took a quick look over the mocks changes and nothing stood out to me as an issue. Just a quick note (see my comment) about keeping an eye out for any data that might be displayed on screen.

Thanks for working on these updates! 🤩

pachlava committed:
Reverted unnecessary user-associated website address change.

pachlava commented:
Thanks for raising the concern about `automatticwidgets.com` ~> `automatticwidgets.wpcomstaging.com` change @rachelmcr and @mokagio! It _looks_ like the website mentioned in the `site-info-wordpress-com.json` [here](https://github.com/woocommerce/woocommerce-android/blob/f398c301f67640ade757be4939757dbfeceece4b/WooCommerce/src/androidTest/assets/mocks/mappings/sites/site-info-wordpress-com.json#L15) will affect the rest of the flow. Initially I had the URL change present in this file. After undoing it, I could undo the change in all other files. Now the PR is much less noisy.

mokagio reviewed (commented):
> Use the `gradle.properties` from secrets

With "secrets", do you me our internal `mobile-secrets` repo? If so, that didn't work for me 😞 

The tests logged in successfully once I updated `gradle.properties` to use:

```diff
- wc.screenshots.url = automatticwidgets.com	
+ wc.screenshots.url = automatticwidgets.wpcomstaging.com
```

I guess we should update the value upstream in `mobile-secrets`. Let me know if you need help with that.

pachlava commented:
> I guess we should update the value upstream in mobile-secrets. Let me know if you need help with that.

I think this should be done. Generally, all references to real resources, like images, will work only after adding the `wpcomstaging.` part. It looks like the real URL has changed (plan expired, or other reasons 🤷). 

So, for consistency, it would be nice to have it updated. I'd be grateful if you could do this @mokagio, because I'm worried I might screw things up there! Thank you! (no urgency with this, though)

pachlava committed:
Disabled WireMock bumps by dependabot

mokagio committed:
Update `gradle.properties` secret file with new screenshots store URL

See
https://github.com/woocommerce/woocommerce-android/pull/4966#pullrequestreview-776795462

mokagio commented:
> I'd be grateful if you could do this @mokagio

Done in [`ded07b5`](https://github.com/woocommerce/woocommerce-android/pull/4966/commits/ded07b56167d56cffb5d9c65625911b5809dc15f)

By the way, it's good to worry about making mistakes because it helps staying on our toes (I'm super paranoid, which is likely the reason I like automated testing this much: it gives me confidence). But, it's also cool to try things out and make a mess, there's always something to learn from it.

AliSoftware commented:
> By the way, it's good to worry about making mistakes because it helps staying on our toes (I'm super paranoid, which is likely the reason I like automated testing this much: it gives me confidence). But, it's also cool to try things out and make a mess, there's always something to learn from it.

👏  well said!

pachlava commented:
Thank you for the review and mobile-secrets help @mokagio !

> But, it's also cool to try things out and make a mess, there's always something to learn from it.

I agree. Probably the case with `mobile-secrets` was an example where I became too afraid / unfamiliar with the environment, and this stopped me.
